### DISCUSSION TITLE
V3: Upgrade Guide additions

### DISCUSSION THREAD

**Discussion by renky** [livewire/livewire#5840] [4. General]
I just upgraded my application (and I'm still not finished) and want to give some hints that should be part of upgrade guide - maybe in an advanced section

### Livewire internal Events changed
I made use of Livewire internal events like 'component.mount'
This was used to get a hook that is fired AFTER all traits have been mounted and was done like this:
```
Livewire::listen('component.mount', fn.....)
```
Currently I don't have a solution for the exact same behaviour - but maybe it can just be listened to the 'mount' event because I hope that the last registered eventlisteners will also be called last.

### Property $attributes on Components must be renamed.
I made heavy use of a property called "attributes" on my Components, that always has been an array. This results in an Error because v3 Components now have a own property with that name. Finally I renamed my properties to $attr now.

### Typehinting for Redirector - Livewire\Redirector removed
I used typehinting for my components to make PHPStan happy. Thats why methods have returnvalues like Livewire\Redirector or combinations of it (in combination with redirect()-Helper Function it must be Livewire\Redirector|Illuminate\Http\RedirectResponse
Livewire redirects now return type Livewire\Features\SupportRedirects\Redirector

### Uppercase Livewire-Object
In JS: window.livewire changed to window.Livewire

### Wireables must be array
in v2 toLivewire could be used with scalar-values - like this:
```
public function toLivewire(): string
{
    return $this::class;
}

public static function fromLivewire(mixed $value): self
{
    return new $value();
}
```

In v3 it must be an array
```
public function toLivewire(): array
{
    return ['className' => $this::class];
}
public static function fromLivewire(mixed $value): self
{
    return new $value['className']();
}
```

### Additional Info for new event dispatching -> is is not just renaming !!
in V2 this was possible:
```
JS:
livewire.emit('event', 'payload1', 'payload2' ....);

PHP
public function listener($payload1, $payload2....);
```

in V3 it must be changed
```
JS:
livewire.dispatch('event', ['payload1', 'payload2',...]);
```
while PHP can stay as it is.

On the other hand it in v2 this was possible
```
PHP 
$this->dispatchBrowserEvent('event', [$payload1, $payload2...])

JS/BLADE
... @event="console.log($event.detail)
where detail was an array[payload1, payload2...]
```

in v3 this must be changed on PHP-Side
```
$this->dispatch('event', $payload1, $payload2, ....);
```
while JS/BLADE can stay as it is

**Comment by joshhanley** [comment]
@renky thanks for this! Any chance you could submit a PR updating the upgrade guide? Thanks!
